## NLP Entity Analysis of Trump Tweets

This is part of my Trump Tweet Analysis project that analyzed tweets between May 2009 and September 2017. 

I have used SpaCy to conduct an enttity analysis on these tweets.  

I then visualized the entity analysis in Tableau. A video overview of this visualization is available at https://youtu.be/mml56SAz2sk

Import the required packages, including both spacy and nlp wth the english dictionary.

In [2]:
import spacy 
nlp = spacy.load('en')
import dill
import os
import pandas as pd
os.chdir('/Users/donajstewart/Master_JSON_Files')


I previously created the merged_df dataframe from JSON files each containing 1 year of tweets, dropped columns I did not need, and pickled it. There are 31,935 tweets in the dataset.

In [3]:
merged_df=dill.load(open('merged_df.pkg', 'rb')) 
         

In [4]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 31935 entries, 0 to 1742
Data columns (total 5 columns):
id_str            31935 non-null int64
created_at        31935 non-null datetime64[ns]
retweet_count     31935 non-null int64
favorite_count    31935 non-null int64
text              31935 non-null object
dtypes: datetime64[ns](1), int64(3), object(1)
memory usage: 1.5+ MB


The dataframe contains the unique id_string for each tweet, the retweet coun, favorite count and the text of the tweet.

In [5]:
merged_df.head(10)

id_str          created_at  retweet_count  favorite_count  \
0  6971079756 2009-12-23 17:38:18             28              12   
1  6312794445 2009-12-03 19:39:09             33               6   
2  6090839867 2009-11-26 19:55:38             13              11   
3  5775731054 2009-11-16 21:06:10              5               3   
4  5364614040 2009-11-02 14:57:56              7               6   
5  5203117820 2009-10-27 15:31:48              4               5   
6  5069623974 2009-10-22 13:57:04              2               2   
7  4862580190 2009-10-14 14:13:17              4              10   
8  4629116949 2009-10-05 14:37:38              1               4   
9  4472353826 2009-09-29 15:28:23             23              30   

                                                text  
0  From Donald Trump: Wishing everyone a wonderfu...  
1  Trump International Tower in Chicago ranked 6t...  
2  Wishing you and yours a very Happy and Bountif...  
3  Donald Trump Partners with TV1 on New Reality ...  
4  --Work has begun, ahead of schedule, to build ...  
5  --From Donald Trump: "Ivanka and Jared’s weddi...  
6  Hear Donald Trump discuss big gov spending, ba...  
7  Watch video of Ivanka Trump sharing business a...  
8  - Read what Donald Trump has to say about daug...  
9  "A lot of people have imagination, but can't e...

## Part of Speech Tagging

We can interate through the text column and create parts of speech tags for each word.  NNP = proper noun, IN = conjunction, VB = verb.  A list of SpaCy annotation can be found here: https://spacy.io/api/annotation

In [6]:
for tweets_doc in nlp.pipe(iter(merged_df['text']), batch_size=1, n_threads=4):
...     print (tweets_doc[0].text, tweets_doc[0].tag_)
...       
           

From IN
Trump NNP
Wishing VBG
Donald NNP
--Work NN
--From IN
Hear VB
Watch VB
- :
" ``
Read VB
- :
- :
Reminder NN
Watch VB
-- :
Ivanka NNP
" ``
Browse NNP
Check VB
Congrats NNS
Donald NNP
" ``
Here RB
Watch VB
RE NN
Donald NNP
“ NFP
- :
RE NN
Thanks NNS
Today NN
Last JJ
“ NFP
Check VB
" ``
Read VB
" ``
Did VBD
Do VB
" ``
Read VB
" ``
" ``
" ``
" ``
" ``
Enter VB
" ``
Listen VB
Miss NNP
" ``
New JJ
Donald NNP
Donald NNP
Be VB
WIshing VBG
Wishing VBG
Do VB
I PRP
... :
Those DT
All PDT
Congratulations NNS
Tonight NN
Tonight NN
Congratulations NNS
My PRP$
Tomorrow NN
Watch VB
Staff NNP
" ``
We PRP
My PRP$
Tonight NN
Tonight NN
I PRP
I PRP
Be VB
Do VB
An DT
Be VB
Scotland NNP
Do VB
I PRP
Read VB
Tune NN
See VB
Do VB
Be VB
Do VB
Eric NNP
I PRP
Coming VBG
There EX
I PRP
Check VB
Check VB
I PRP
Went VBD
Spent NNP
The DT
Congratulations NNS
It PRP
Four CD
The DT
So RB
That DT
Friday NNP
Check VB
Check VB
Eric NNP
Mark NNP
Performing VBG
The DT
Enter VB
and CC
I PRP
Could MD
I PRP
and CC
and CC

## Named Entities

We can also create a list of entities - which differs from parts of speech.  Named Entities include persons, organizations, and geographical locations, see: https://spacy.io/api/annotation#named-entities   We need to load the NLP pipeline.

In [7]:
tweetentities = []
for tweets_doc in nlp.pipe(iter(merged_df['text']), batch_size=1, n_threads=4):
    for ent in tweets_doc.ents:
        tweetentities.append([ent.label_,ent.text]) 
print(tweetentities) 


[['PERSON', 'Donald Trump'], ['EVENT', 'New Year'], ['DATE', '2010'], ['ORG', 'Trump International Tower'], ['GPE', 'Chicago'], ['ORDINAL', '6th'], ['ORG', 'Council'], ['ORG', 'Tall Buildings & Urban Habitat'], ['WORK_OF_ART', 'Happy and Bountiful Thanksgiving'], ['PERSON', 'Donald Trump Partners'], ['EVENT', 'New Reality Series Entitled'], ['ORG', 'Omarosa'], ['ORG', 'Ultimate Merger'], ['ORG', 'Trump International – Scotland'], ['PERSON', 'Donald Trump'], ['WORK_OF_ART', 'Ivanka and Jared’s'], ['PERSON', 'Donald Trump'], ['EVENT', 'World w'], ['PERSON', 'Ivanka Trump'], ['CARDINAL', '4'], ['ORG', 'GMA'], ['PERSON', 'Donald Trump'], ['ORG', 'Ivanka'], ['ORG', 'The Trump Card'], ['PERSON', 'J. Trump'], ['PERSON', "Donald Trump's"], ['WORK_OF_ART', 'Ten Tips for Success: http://tinyurl.com/mcn8yl'], ['ORG', 'DSRL'], ['PERSON', 'Donald Trump'], ['WORK_OF_ART', 'Double Trump'], ['ORG', 'Golden Lick Race Sweepstakes'], ['PERSON', 'Donald Trump'], ['WORK_OF_ART', 'The Miss Universe'], ['GPE

We can place these entities in a new datafarame for further analysis using pandas.

In [8]:
tweetents_df=pd.DataFrame(tweetentities, columns={'entity', 'text'})
tweetents_df.head(5)

entity                       text
0  PERSON               Donald Trump
1   EVENT                   New Year
2    DATE                       2010
3     ORG  Trump International Tower
4     GPE                    Chicago

We now have 52,458 entities extracted from the tweets.

In [9]:
tweetents_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52458 entries, 0 to 52457
Data columns (total 2 columns):
entity    52458 non-null object
text      52458 non-null object
dtypes: object(2)
memory usage: 819.7+ KB


The entity list can also be saved to csv

In [10]:
tweetents_df.to_csv('tweetents.csv', index=False)

#df.to_csv('example.csv', index=False)

Using groupby we see that PERSON entities account for the most entities, folowed by organizations and persons.

In [11]:

tweetents_df.groupby('entity').size().sort_values(ascending=False) \
  .reset_index(name='text')

entity   text
0        PERSON  13579
1           ORG  11281
2           GPE   9176
3          DATE   4935
4      CARDINAL   3416
5          TIME   2158
6   WORK_OF_ART   2072
7          NORP   1991
8         MONEY   1288
9       ORDINAL    660
10      PERCENT    593
11          LOC    365
12          FAC    330
13        EVENT    268
14     QUANTITY    153
15          LAW     85
16      PRODUCT     82
17     LANGUAGE     26

We can also see that the word Trump and Donald Trump are the most common values in the text column, followed by America.

In [12]:

tweetents_df['text'].value_counts()

Trump                                         1827
Donald Trump                                   873
America                                        830
Obama                                          809
today                                          619
tonight                                        583
one                                            528
China                                          477
U.S.                                           425
2016                                           416
Donald                                         384
ObamaCare                                      349
American                                       341
first                                          321
tomorrow                                       261
Hillary                                        251
US                                             245
TRUMP                                          227
Iran                                           213
Republicans                    

Among the Organizzations identified are 'Trump International Tower", 'Ivanka', and 'Tall Buidings & Urban Habitat', illustrating some of the limitations of NLP analysis. Overall there are 11,281 ORG entities.

In [13]:
tweetents_df.loc[tweetents_df['entity'] == 'ORG'] 

entity                                               text
3        ORG                          Trump International Tower
6        ORG                                            Council
7        ORG                     Tall Buildings & Urban Habitat
11       ORG                                            Omarosa
12       ORG                                    Ultimate Merger
13       ORG                     Trump International – Scotland
20       ORG                                                GMA
22       ORG                                             Ivanka
23       ORG                                     The Trump Card
27       ORG                                               DSRL
30       ORG                       Golden Lick Race Sweepstakes
36       ORG                                                NBC
38       ORG                                               LIVE
42       ORG                                                NBC
49       ORG                                Summer Reading List
50       ORG                          the Trump University Blog
53       ORG                          the Trump University Blog
57       ORG                           N.J. Lieutenant Governor
70       ORG                                     FB Vanity URLs
71       ORG                                   SF Chronicle - "
75       ORG                                   Britney Spears &
82       ORG                                   iGoogle Showcase
93       ORG                        The New York Times Magazine
104      ORG                                    USA Tara Conner
108      ORG                        Celebrity Apprentice Finale
110      ORG                                 Ten Financial Tips
125      ORG                                                CNN
131      ORG                                                CNN
134      ORG                          Trump International Hotel
136      ORG  Trump International Hotel & Tower Waikiki Beac...
...      ...                                                ...
52339    ORG                                    Conde Nast &amp
52341    ORG                                                 AM
52345    ORG                                        Kasich &amp
52348    ORG                                                DJT
52352    ORG                                              SAW A
52358    ORG                  The Democratic National Committee
52359    ORG                                                FBI
52361    ORG                                                NBC
52362    ORG                                              Obama
52363    ORG                                       Toyota Motor
52374    ORG                                              BLAME
52375    ORG                                          ObamaCare
52380    ORG                                      the "Movement
52381    ORG                                          ObamaCare
52389    ORG                                          ObamaCare
52390    ORG                                                DNC
52391    ORG                                                RNC
52392    ORG                                               Ford
52408    ORG                                            AMERICA
52409    ORG                                    INNOVATION &amp
52415    ORG                                           Congress
52416    ORG                    the Independent Ethics Watchdog
52419    ORG                            The Affordable Care Act
52420    ORG                                          ObamaCare
52421    ORG                                          ObamaCare
52425    ORG                                              CRAZY
52426    ORG                                     General Motors
52447    ORG                                       MAKE AMERICA
52450    ORG                                          Happy New
52456    ORG                                         POTUS &amp

[11281 rows x 2 columns]

NORP stands for nationalities, religious or political groups. There are 52,527 NORP entities.

In [14]:
tweetents_df.loc[tweetents_df['entity'] == 'NORP']

entity               text
507     NORP           Scottish
558     NORP           American
600     NORP          Americans
601     NORP  Indy Presidential
608     NORP             Iraqis
656     NORP          Americans
661     NORP             Iraqis
672     NORP           American
674     NORP        Republicans
735     NORP           European
740     NORP       @JonHuntsman
778     NORP          Americans
794     NORP           Iraqis--
819     NORP            Chinese
889     NORP           Islamist
910     NORP          Americans
915     NORP          Americans
917     NORP          Americans
955     NORP           European
976     NORP           American
988     NORP        politicians
1012    NORP           American
1016    NORP            Iranian
1049    NORP        Republicans
1093    NORP        Republicans
1100    NORP          Americans
1105    NORP          Americans
1121    NORP               Arab
1144    NORP            Chinese
1146    NORP        Republicans
...      ...                ...
51817   NORP           Canadian
51868   NORP           Japanese
51907   NORP           American
51918   NORP          Democrats
51928   NORP            Islamic
51964   NORP            Islamic
52006   NORP          Democrats
52011   NORP          Democrats
52013   NORP           American
52014   NORP          Democrats
52033   NORP         Christians
52083   NORP           American
52104   NORP               Feds
52139   NORP          Americans
52146   NORP           American
52163   NORP           American
52223   NORP            Russian
52236   NORP          Democrats
52237   NORP    Obama Democrats
52254   NORP          Democrats
52255   NORP        Republicans
52372   NORP        Republicans
52373   NORP          Democrats
52376   NORP          Democrats
52383   NORP               Dems
52387   NORP        Republicans
52401   NORP           Russians
52403   NORP            Russian
52417   NORP           Democrat
52427   NORP            Mexican

[1991 rows x 2 columns]

GPE includes countries and other places. Again, however, it is not perfect - Melania and Facebook both show up as places.

In [15]:
tweetents_df.loc[tweetents_df['entity'] == 'GPE'] 

entity          text
4        GPE       Chicago
33       GPE       Bahamas
88       GPE      Facebook
96       GPE    Trumpative
135      GPE     Las Vegas
148      GPE    Apprentice
216      GPE      Scotland
223      GPE      Aberdeen
224      GPE      Scotland
236      GPE       Melania
260      GPE          U.S.
278      GPE        Mexico
282      GPE     Las Vegas
319      GPE     Las Vegas
321      GPE              
324      GPE     Las Vegas
333      GPE      New York
346      GPE      Scotland
359      GPE       Melania
360      GPE        Ivanka
372      GPE      New York
406      GPE      Scotland
430      GPE           USA
433      GPE           USA
440      GPE     Las Vegas
470      GPE     Las Vegas
489      GPE    Palm Beach
500      GPE      New York
513      GPE  Central Park
546      GPE    Palm Beach
...      ...           ...
52326    GPE        Russia
52327    GPE        Russia
52332    GPE       Florida
52349    GPE          D.C.
52357    GPE        Mexico
52360    GPE        Russia
52364    GPE          Baja
52365    GPE        Mexico
52367    GPE          U.S.
52368    GPE          U.S.
52370    GPE     Agreement
52385    GPE       Arizona
52393    GPE        Mexico
52395    GPE          U.S.
52397    GPE          U.S.
52406    GPE       America
52407    GPE         Gitmo
52411    GPE        Mexico
52412    GPE      Michigan
52423    GPE       Arizona
52428    GPE   Chevy Cruze
52429    GPE          U.S.
52430    GPE      U.S.A.or
52431    GPE         China
52432    GPE          U.S.
52434    GPE   North Korea
52435    GPE   North Korea
52436    GPE          U.S.
52441    GPE       Chicago
52455    GPE       America

[9176 rows x 2 columns]

To visualize this data, and perform additional analysis, I imported it into Tableau. Video walkthrough available at https://youtu.be/mml56SAz2sk